### ISSUE TITLE
[Feature Request] The role of `observation_spec` and `output_spec`

### ISSUE THREAD

**Issue by matteobettini** [pytorch/rl#1068]
## The current state of things

Currently, the output of an env is described by an `output_spec`
```
output_spec{
     done -> done_spec
     reward -> reward_spec
     observation -> observation_spec{
               observation -> real_observation_spec
     }
}
```

## What is the problem?

In my opinion this might be confusing because:
1. The observation spec must be composite and have an observation key
2.  The `output_spec` is just an container of predetermined specs. In fact, if you look at https://github.com/pytorch/rl/blob/main/torchrl/envs/common.py#L791, you can see that even if you added new keys to output specs, these are not considered in the fake_tensordict, giving errors in the code when those keys will be present in tensordicts and not expected.

As a result of 2 (but also as a choice in the library design) all extra info coming from the environment MUST be placed in `observation_spec`. 

For example, your env may output infos like `bitcoin_cost_of_action_taken` which you may not want to put under `observation_spec` as it should not be an observation, but just a metric.

### Option 1
Just fixing issue 2 by enabling to add other keys to `output_spec` which are infos that are not observations.

### Option 2

To fix 1 and 2 we can think of this:

Since we now enforce `observation_spec` to be composite and have an `observation` key, why not enforcing directly `output_spec` to be composite (which is already) and have an `observation_spec` (in the same way as reward and done) which can be any spec. 

Therefore, users can decide where to put their infos.

For example let's say that my env outputs 3 keys: `stock_price`, `bitcoin_cost_of_action_taken` and `number_of_stocks_sold`.

Normally, I would have to put them all under the `observation_spec` and decide which one to call `observation`. But altrenatively I could act in the following ways:

- Case 1: I want my observation to be `stock_price` and the rest to be infos. Then I create an `observation_spec` which is non composite with the bounds of `stock_price` and I return tensordicts with the "observation" key and the value corresponding to `stock_price` . I put `bitcoin_cost_of_action_taken` and `number_of_stocks_sold` as keys in `output_spec`.
- Case 2: I want my observation to be `stock_price` and `number_of_stocks_sold` and the rest to be infos. Then I create an `observation_spec` which is composite and has keys `stock_price` and `number_of_stocks_sold` and I return  tensordicts with these keys and their values. I put `bitcoin_cost_of_action_taken` in output_spec.

In this solution, to check what are my observations I would look at `observation_spec`. 

- If it is non composite I know that my observation is simply the one under the "observation" key in a tensordict
- If it is composite I will look for the respective keys  in the tensordict (this still allows to have one key named "observation" but this could also not be the case)

**Comment by vmoens** [comment]
> Since we now enforce observation_spec to be composite and have an observation key, why not enforcing directly output_spec to be composite (which is already) and have an observation_spec (in the same way as reward and done) which can be any spec.

This is not possible for the following reason:
it is common to have an env that returns a single obs, say
```
env = GymEnv("Pendulum-v1")
```
Now following this logic this should have
```
output_spec = {"observation": BoundedSpec(...), "done": ..., "reward": ...}
```
Now say I want to render my env in the same script. I have
```
env = GymEnv("Pendulum-v1", from_pixels=True)
```
with
```
output_spec = {"observation": {"state": BoundedSpec(...), "pixels": ...}, "done": ..., "reward": ...}
```
This becomes very confusing and impractical: I need to tell my actor what to do when the env returns pixels and when it doesn't, which is peculiar to say the least.
There are many similar cases but the idea is essentially that we need observation to be a composite and to unpack it. If you have another way of making the above example work seamlessly (ie without headache for unexperienced users) and that is more elegant i'm glad to read it!

**Comment by vmoens** [comment]
For the rest I'm happy with having anything custom in `output_spec`, we can fix `fake_tensordict` to accomodate that.

**Comment by matteobettini** [comment]
Thanks for writing that down, I remembered that we discussed it but I had forgotten the specific example.

we could have what you wrote above and just instead of calling the key “state” in your example we would call it “observation” this would let the actor know that that is the key that corresponds to the uncomposite obs spec in the normal case

So that the only thing that from_pixel does is extending the observation spec by making it composite, putting the observation in the observation key and adding a new key “pixels”.

this is for the case that pixels are for input to the neural network. for just rendering, pixels would be added to the output spec and the obs spec left untouched

**Comment by vmoens** [comment]
but we don't make the observation spec private!
Here just because you call `from_pixels=True` you change the nature of your
spec, that does not seem right to me, especially if on the other side your
input/output structure is unchanged. It becomes quite counter intuitive IMO

**Comment by matteobettini** [comment]
I didn’t get the point about privacy.

But doesn’t `from_pixel=True` change the input/output? It adds a new pixel env output for the neural network to use, so there are more observations now cause you have the pixels too.

 if the pixels are only for rendering and are not to be considered an obs, that key can just be added to the output spec

Btw, i am also down to just go with solution 1 like you said. I am just poking to see if we can brainstorm an idea for solution 2

**Comment by matteobettini** [comment]
Another option is to treat observation_spec like the one that is give and never modify it and operations like from_pixels always add stuff just to the output_spec.

**Comment by vmoens** [comment]
> Another option is to treat observation_spec like the one that is given and never modify it and operations like from_pixels always add stuff just to the output_spec.

I think that would be very surprising to multiple users. What is a pixel if not a an observation? In atari (always from pixels), you would put the pixels out of the observation spec? Or you would have pixels sometimes there sometimes somewhere else? How would `ToTensorImage` know where to fetch the pixel spec to transform (in output or observation)? Seems like a lot of headaches to me.

> aka observation_spec just becomes the content of observaton_spec[“observation”] and the rest goes in output_spec
> 
> this follows your philosophy even more as we do no modifications at all to observation_spec with from_pixel
> 
> users will know that if they ask for extra pixels they will find them in output_spec[“pixels”] and they can decides what parts of output_spec to add to their nn input according to their liking

I disagree, I think user won't "know" anything and will be surprised to find pixels in output and not observation. Most user won't care about `env.output_spec` and will only look at `env.observation_spec` IMO.

What I still don't get is what is the problem of having extra stuff in observation spec? Why do we need extra levels of complexity?

It seems to me that if we start saying "we put step_count in output_spec, info in output_spec["info"], the trajectory id in output_spec["traj_id"], the pixels in output_spec["observation", "pixels"]..." we're going to have to document a lot of opinionated choices that will scare people off.
IMO it's much easier to just say "outpu_spec contains the done, reward and observation. observation contains everything else." It's a one liner that is loosely restrictive (since you can put whatever you want in observation spec anyway).

I think the confusion is just about the term "observation". If that is the problem let's rename it data or give an alias with whichever name the community thinks is fit. But a naming issue should not add extra levels of complexity to an env class that is already hard ot maintain in its current state...

**Comment by matteobettini** [comment]
Ok, well it was good to write down this conversation anyway so maybe in the future users can weigh in. 

From a retrocompatibility of user mindsets I think having both infos and observations in the observation spec is surprising and the same goes for the forced observation key.

About solution 1, is it ok if I make a PR to allow extra keys in output_spec or now we want to keep it forcibly only contain reward, done and obs? 
Cause now the fake_tensordict function causes bugs when users add more keys to their output spec

**Comment by vmoens** [comment]
I agree it's good to have these discussions
My TLDR is:
- I agree that the current state is suboptimal and could be improved
- I agree that the "observation" name is confusing
- Any solution would need to satisfy these items:
  - it must be bc-compatible
  - it must be simple: `GymEnv(..., from_pixels=True)` `GymEnv(..., from_pixels=False)` must have the same spec structure, and the same output structure (we can't have the state suddenly moved in a nested dict)
  - I don't think it is wise to always have states written in `tensordict["next", "observation", "pixels"]` or smth like that (this is what the expansion of `observation_spec` at the root level is supposed to address) because that would be unintuivite for many users. The "basic" user will wonder why a single state is nested three levels down in the tensordict.
  -  the solution needs to be transparent: if "pixels" is in observation_spec, and I apply a transform to these pixels, the resulting entry should still be in observation_spec. In general Transforms should not be more complex if we refactor this (ie I'm not in favour of `CatTensors(..., spec_dest="output")` bc it's going to be a nightmare to maintain).


Things I'm not sure about:
- I decided to put the StepCounter output in observation_spec. It is not an observation, but if we start to say "step_count goes in output_spec, infos go in output, transforms of obs go in observation, ..." i'm afraid that we'll end up to a place where we need to make quite opinionated choices that will require our used to dig in the doc to find the relevant information. Putting everything in observation_spec is convenient. Example: for decision transformers we cat "obs" "action" and "reward". This creates a new item to feed to the policy. Where should its spec go? Bc if I stack "joint-position" and "joint-speed" from a robot, I expect this tensor to live in observation spec. So what is the rule to put the result somewhere else, and more importantly, how can we make that transparent?
```
env.append_transform(CatTensors(in_keys=["action", "obs", "reward"], out_key=["new_obs"])) # where should the "new_obs" spec go? 
env.append_transform(CatTensors(in_keys=["join-pos", "joint-speed"], out_key=["new_obs"])) # where should the "new_obs" spec go? 
```
-  Same for info-keys. I understant that these are not obs, but how can we be transparent about why we put them somewhere else? What if they are used as state for training?

**Comment by vmoens** [comment]
I don't want to give the impression that i'm against refactoring this, actually i don't like it either
it's been tickling me for quite some time
but IDK what to do that wouldn't break things apart or make the UX of basic users worse
